csrc/scheduler/matmul.cpp

```diff
@@ -1304,8 +1305,21 @@ void scheduleMatmul(Fusion* fusion, const MatmulParams& params) {
   }
 
   if (params.circular_buffer_options.circular_buffer_smem_read) {
-    acr->circularBuffer(/*number_of_stages=*/2);
-    bcr->circularBuffer(/*number_of_stages=*/2);
+    constexpr int64_t number_of_stages = 2;
+
+    IterDomain* acr_axis = getCircularBufferAxis(acr);
+    NVF_ERROR(acr_axis != nullptr);
+    if (acr_axis->extent()->isConstScalar() &&
+        acr_axis->extent()->evaluate() >= number_of_stages) {
+      acr->circularBuffer(number_of_stages);
+    }
+
+    IterDomain* bcr_axis = getCircularBufferAxis(bcr);
+    NVF_ERROR(bcr_axis != nullptr);
+    if (bcr_axis->extent()->isConstScalar() &&
+        bcr_axis->extent()->evaluate() >= number_of_stages) {
+      bcr->circularBuffer(number_of_stages);
+    }
```

Comment:
Nit: you could create a lambda to avoid repeating this part.

Reply:
Instead of an if statement, should we just assert that these are constant? That could be done by just removing the `isConstScalar` check in the condition.